### DISCUSSION TITLE
How to build it on Apple M1?

### DISCUSSION THREAD

**Discussion by evstarshov** [LostRuins/koboldcpp#121] [Q&A] [answered]
Works fine with my intel mac, but on M1 i got these errors when I launch make file:

clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
otherarch/ggml_v1.c:1152:21: error: always_inline function '_mm256_cvtph_ps' requires target feature 'f16c', but would be inlined into function 'ggml_v1_vec_dot_f16' that is compiled without support for 'f16c'
            ax[j] = GGML_V1_F16_VEC_LOAD(x + i + j*GGML_V1_F16_EPR, j);
                    ^
otherarch/ggml_v1.c:800:40: note: expanded from macro 'GGML_V1_F16_VEC_LOAD'
#define GGML_V1_F16_VEC_LOAD(p, i)     GGML_V1_F32Cx8_LOAD(p)
                                       ^
otherarch/ggml_v1.c:790:36: note: expanded from macro 'GGML_V1_F32Cx8_LOAD'
#define GGML_V1_F32Cx8_LOAD(x)     _mm256_cvtph_ps(_mm_loadu_si128((__m128i *)(x)))
                                   ^
otherarch/ggml_v1.c:1153:21: error: always_inline function '_mm256_cvtph_ps' requires target feature 'f16c', but would be inlined into function 'ggml_v1_vec_dot_f16' that is compiled without support for 'f16c'
            ay[j] = GGML_V1_F16_VEC_LOAD(y + i + j*GGML_V1_F16_EPR, j);
                    ^
otherarch/ggml_v1.c:800:40: note: expanded from macro 'GGML_V1_F16_VEC_LOAD'
#define GGML_V1_F16_VEC_LOAD(p, i)     GGML_V1_F32Cx8_LOAD(p)
                                       ^
otherarch/ggml_v1.c:790:36: note: expanded from macro 'GGML_V1_F32Cx8_LOAD'
#define GGML_V1_F32Cx8_LOAD(x)     _mm256_cvtph_ps(_mm_loadu_si128((__m128i *)(x)))
                                   ^
otherarch/ggml_v1.c:1530:21: error: always_inline function '_mm256_cvtph_ps' requires target feature 'f16c', but would be inlined into function 'ggml_v1_vec_mad_f16' that is compiled without support for 'f16c'
            ax[j] = GGML_V1_F16_VEC_LOAD(x + i + j*GGML_V1_F16_EPR, j);
                    ^
otherarch/ggml_v1.c:800:40: note: expanded from macro 'GGML_V1_F16_VEC_LOAD'
#define GGML_V1_F16_VEC_LOAD(p, i)     GGML_V1_F32Cx8_LOAD(p)
                                       ^
otherarch/ggml_v1.c:790:36: note: expanded from macro 'GGML_V1_F32Cx8_LOAD'
#define GGML_V1_F32Cx8_LOAD(x)     _mm256_cvtph_ps(_mm_loadu_si128((__m128i *)(x)))
                                   ^
otherarch/ggml_v1.c:1531:21: error: always_inline function '_mm256_cvtph_ps' requires target feature 'f16c', but would be inlined into function 'ggml_v1_vec_mad_f16' that is compiled without support for 'f16c'
            ay[j] = GGML_V1_F16_VEC_LOAD(y + i + j*GGML_V1_F16_EPR, j);
                    ^

**Comment by LostRuins** [comment]
Please try the latest commit and see if the issue is fixed.

**Reply by evstarshov** [reply]
It compiled, thank you, but when I launch start script i got this error:

OSError: dlopen(/Volumes/KINGSTON/M1/koboldcpp/koboldcpp.so, 0x0006): tried: '/Volumes/KINGSTON/M1/koboldcpp/koboldcpp.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64')), '/System/Volumes/Preboot/Cryptexes/OS/Volumes/KINGSTON/M1/koboldcpp/koboldcpp.so' (no such file), '/Volumes/KINGSTON/M1/koboldcpp/koboldcpp.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64'))

**Reply by LostRuins** [reply]
What did you use to build it? Sounds like you used an x86 compiler, but your M1 is running an ARM based chip.

**Reply by evstarshov** [reply]
Only make command inside without any options. This reproduces even after re-cloning repository to another folder

**Reply by LostRuins** [reply]
Yeah I mean, the compiler you used. It seems like you are trying to build for the wrong architecture?

**Reply by evstarshov** [reply]
Can you give me advice how to change compiler? I tried to add to Makefile line 174: -arch arm64. It compiles, but after start window and selecting model I have another error:

Welcome to KoboldCpp - Version 1.15
For command line arguments, please refer to --help
Otherwise, please manually select ggml file:
2023-04-28 12:56:09.007 python3[22414:754319] +[CATransaction synchronize] called within transaction
Warning: OpenBLAS library file not found. Non-BLAS library will be used.
Initializing dynamic library: koboldcpp.so
Traceback (most recent call last):
  File "/Volumes/KINGSTON/M1/koboldcpp/koboldcpp.py", line 588, in <module>
    main(args)
  File "/Volumes/KINGSTON/M1/koboldcpp/koboldcpp.py", line 499, in main
    init_library() # Note: if blas does not exist and is enabled, program will crash. 
  File "/Volumes/KINGSTON/M1/koboldcpp/koboldcpp.py", line 122, in init_library
    handle.load_model.argtypes = [load_model_inputs] 
  File "/Users/evgenijstarsov/miniconda3/lib/python3.10/ctypes/__init__.py", line 387, in __getattr__
    func = self.__getitem__(name)
  File "/Users/evgenijstarsov/miniconda3/lib/python3.10/ctypes/__init__.py", line 392, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: dlsym(0x85df4bf0, load_model): symbol not found

**Reply by Myrminki** [reply]
Holy cow, now I can compile it on my old xeon e5 1620 0 on Linux, the other versions gave me similar errors, so I tried again and now it works.

**Reply by LostRuins** [reply]
nice

**Reply by evstarshov** [reply, marked as answer]
I found my problem why it won't work on M1 mac after compiling - my terminal was opened with enabled Rosetta option. I enabled it to run another project and forgot about it. Now everything compiles and works perfectly.
